plugins/jenkins-backend/src/service/jenkinsInfoProvider.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  Entity,
+  EntityName,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
+import { CatalogClient } from '@backstage/catalog-client';
+import { Config } from '@backstage/config';
+
+export interface JenkinsInfoProvider {
+  getInstance(options: {
+    /**
+     * The entity to get the info about.
+     */
+    entityRef: EntityName;
+    /**
+     * A specific job to get. This is only passed in when we know about a job name we are interested in.
+     */
+    jobName?: string;
+  }): Promise<JenkinsInfo>;
+}
+
+export interface JenkinsInfo {
+  baseUrl: string;
+  headers?: Record<string, string | string[]>;
+  jobName: string; // TODO: make this an array
+}
+
+/**
+ * Use default config and annotations.
+ *
+ * This will fallback through various deprecated config and annotation schemes.
+ */
+export class DefaultJenkinsInfoProvider implements JenkinsInfoProvider {
+  static readonly OLD_JENKINS_ANNOTATION = 'jenkins.io/github-folder';
+  static readonly NEW_JENKINS_ANNOTATION = 'jenkins.io/job-slug';
+
+  constructor(
+    private readonly catalog: CatalogClient,
+    private readonly config: Config,
```

Comment:
I recommend making this into a `static fromConfig(options: { config: Config, catalog ...}): DefaultJenkinsInfoProvider`.

In there, you would unpack and translate the config into just a plain `{ name: string, baseUrl: string, ... }[]` (but with a named type for the object I guess) which is then given to the constructor of the actual class.

That also removes some of the error handling in `getInstanceConfig` later at runtime. Which is a good thing, because the adopter gets notified about config errors at startup instead of functionality breaking later down the line